File: web3-libs/mod4/playground/aula4/getTxFromBlock.js
// 1. Instale e importe a biblioteca Ethers
import { ethers } from "ethers";
import readline from "readline";

// 2. Configure o provider
const RPC_URL = "http://127.0.0.1:8545";
const provider = new ethers.JsonRpcProvider(RPC_URL);

async function main() {
  console.log("Do you want to consult some transactions?");

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });

  rl.question(
    "Enter block hash (or none for the latest block): ",
    async (answer) => {
      if (answer === "") {
        const currentBlockNumber = await provider.getBlockNumber();
        const block = await provider.getBlock(currentBlockNumber);
        const txCount = block.transactionCount;
        console.log(`This block hash ${txCount} Transaction`);
        
        const transactions = await Promise.all(
          block.transactions.map(txHash => provider.getTransaction(txHash))
        );
        
        console.log(transactions);
      } else {
        // Verifica se o hash do bloco é válido
        if (!/^0x[a-fA-F0-9]{64}$/.test(answer)) {
          console.error("Invalid block hash. Please enter a valid hash.");
          rl.close();
          return;
        }

        const block = await provider.getBlock(answer);
        if (!block) {
          console.error("Block not found. Please check the hash.");
        } else {
          const txCount = block.transactionCount;
          console.log(`This block hash ${txCount} Transaction`);

          // Obtém as transações do bloco
          const transactions = await Promise.all(
            block.transactions.map(txHash => provider.getTransaction(txHash))
          );
          
          console.log(transactions);
        }
      }
      rl.close();
    }
  );
}

main();

File: web3-libs/mod3/playground/aula4/get_tx_from_block.py
from web3 import Web3
from pprint import pprint

# Configure o provider
RPC_URL = "http://127.0.0.1:8545"
web3 = Web3(Web3.HTTPProvider(RPC_URL))

print("Do you want to consult some transactions?")
block_hash = input("Enter block hash (or none for the latest block): ")

if block_hash == "":
    current_block_number = web3.eth.get_block_number()
    tx_count = web3.eth.get_transaction_count(current_block_number)

    print(f"This block hash {tx_count} Transaction")
    transaction = web3.eth.get_transaction_from_block(current_block_number)
    pprint(dict(transaction))
else:
    transaction = web3.eth.get_transaction_by_block(block_hash, 0)
    tx_count = web3.eth.get_block_transaction_count(transaction.blockNumber)
    print(f"This block hash {tx_count} Transaction")
    pprint(dict(transaction))
